src/shamir.rs

```diff
@@ -0,0 +1,364 @@
+            }
+
+            r += share_i.y * x * maybe_denom.unwrap();
+        }
+
+        Ok(r)
+    }
+}
+
+impl<F: Field> Share<F> {
+    /// Adds two shares together.
+    #[allow(dead_code)]
+    fn add(&self, rhs: &Share<F>) -> Result<Self, ShareError> {
+        if self.x == rhs.x {
+            Ok(Share {
+                x: self.x,
+                y: self.y.add(rhs.y),
+            })
+        } else {
+            Err(ShareError::MisalignedShares {
+                lhs: self.x,
+                rhs: rhs.x,
+            })
+        }
+    }
+}
+
+/// Generate coefficients (sampled at random) for an arbitrary polynomial of infinite degree.
+fn gen_polynomial<R: RngCore, F: Field>(mut rng: R) -> impl Iterator<Item = F> {
+    repeat_with(move || {
+        let p: F = F::random(&mut rng);
+        p
+    })
+}
+
+#[cfg(test)]
+// PrimeField macro panics if this attribute is added to `Fp` struct.
+#[allow(clippy::expl_impl_clone_on_copy)]
+mod tests {
+    use crate::shamir::{Error, SecretSharing, ShareError};
+    use ff::PrimeField;
+    use proptest::prelude::*;
+    use rand::rngs::StdRng;
+    use rand::thread_rng;
+    use rand_core::SeedableRng;
+    use std::cmp::max;
+    use std::num::NonZeroU8;
+
+    // Mersenne prime 2^61-1
+    const PRIME: u64 = 2_305_843_009_213_693_951;
+
+    #[derive(PrimeField)]
+    #[PrimeFieldModulus = "2305843009213693951"]
+    #[PrimeFieldGenerator = "7"]
+    #[PrimeFieldReprEndianness = "little"]
+    struct Fp([u64; 1]);
+
+    impl Fp {
+        pub fn new(v: u64) -> Self {
+            Fp([v])
+        }
+
+        pub const fn modulus() -> u64 {
+            PRIME
+        }
+    }
+
+    #[test]
+    fn can_share_8_byte_int() {
+        for k in [2_u8, 5, 10, 254] {
+            for n in k..=k.checked_add(5).unwrap_or(u8::MAX) {
+                let k = NonZeroU8::new(k).unwrap();
+                let n = NonZeroU8::new(n).unwrap();
+                let shamir = SecretSharing::new(k, n).unwrap();
+                let secret = Fp::new(213);
+                let shares = shamir.split(secret, thread_rng());
+
+                assert_eq!(secret, shamir.reconstruct(&shares).unwrap());
+            }
+        }
+    }
+
+    #[test]
+    fn can_add_shares() {
+        fn check_addition(shamir: &SecretSharing, lhs_secret: u64, rhs_secret: u64) {
+            let mut rng = thread_rng();
+
+            let lhs_secret = Fp::new(lhs_secret);
+            let rhs_secret = Fp::new(rhs_secret);
+            let expected = lhs_secret + rhs_secret;
+
+            let lhs_shares = shamir.split(lhs_secret, &mut rng);
+            let rhs_shares = shamir.split(rhs_secret, &mut rng);
+            let sum_shares = lhs_shares
+                .iter()
+                .zip(rhs_shares.iter())
+                .map(|(lhs, rhs)| lhs.add(rhs).unwrap())
+                .collect::<Vec<_>>();
+
+            let sum_secret = shamir
+                .reconstruct(&sum_shares[0..shamir.threshold() as usize])
```

Comment:
I played with const generics a bit here and they seemed to me less ergonomic to use compared to variable `k` and `n` for the following reasons:

* any function that takes `SecretSharing` as an argument, now needs to be parametrized with const generic argument or make an assumption about the specific `K` used in IPA. 
* Dynamic creation of `SecretSharing` with const generic parameter does not seem to be possible (yet), so testing for K = 1,2,3,...255 gets tricky.
* `usize` is the only type Rust supports now to constrain array bounds (as of now) and it does not fit our use case well because there are a lot of invalid `K` values which we can only reject at runtime (0, 1, 256, ...). We need const expressions feature, otherwise someone can just simply crash our program by providing K = 1_000_000. 

Thinking more about it, there seem to be multiple directions we can go:


* Hardcode K = N = 3. Cache Lagrange coefficients and leverage the knowledge that K is small to leverage stack whenever possible.
* Assume K = N and implement secret sharing with no redundancy but support any K within [2, 255] range. Using const generics is still an option, but as described above it is probably less convenient than dynamic `K ` . Caching Lagrange coefficients is still beneficial and can be implemented, however there will be no compile time checks on `K`. 
* Implement Secret Sharing with maximum flexibility and assume K <= N <= 256 (this still feels reasonable). 

I am personally in favour of option 2, but curious to hear what you think

Reply:
I stand corrected here, we actually need k < n to be able to multiply secret shares and recover the product by using `n` shares, so I am inclined to keep the API as is, unless there is a strong objection against it. When (or rather if) we decide to cache LC, we can start rejecting inputs to reconstruct with number of shares < n.

Reply:
resolved?